mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -715,80 +969,50 @@ boolean onTaskExecutorStatusChange(TaskExecutorStatusChange statusChange) {
 
         private boolean handleStatusChange(TaskExecutorReport report) throws IllegalStateException {
             if (availabilityState == null) {
-                availabilityState = from(report);
-                return true;
+                return setAvailabilityState(from(report));
             } else {
-                switch (availabilityState) {
-                    case Pending:
-                        if (report instanceof Available) {
-                            return false;
-                        } else if (report instanceof Occupied) {
-                            throwInvalidTransition(report);
-                        }
-                    case Assigned:
-                        if (report instanceof Available) {
-                            return false;
-                        } else if (report instanceof Occupied) {
-                            if (((Occupied) report).getWorkerId().equals(workerId)) {
-                                this.availabilityState = AvailabilityState.Running;
-                                return true;
-                            } else {
-                                throwInvalidTransition(report);
-                            }
-                        }
-                    case Running:
-                        if (report instanceof Available) {
-                            this.workerId = null;
-                            this.availabilityState = AvailabilityState.Pending;
-                            return true;
-                        } else if (report instanceof Occupied) {
-                            if (!((Occupied) report).getWorkerId().equals(workerId)) {
-                                throwInvalidTransition(report);
-                            } else {
-                                return false;
-                            }
-                        }
+                return setAvailabilityState(availabilityState.onTaskExecutorStatusChange(report));
+            }
+        }
+
+        private boolean setAvailabilityState(AvailabilityState newState) {
+            if (this.availabilityState != newState) {
+                this.availabilityState = newState;
+                if (this.availabilityState instanceof Running) {
+                    if (isDisabled()) {
+                        jobMessageRouter.routeWorkerEvent(new WorkerOnDisabledVM(newState.getWorkerId()));
+                    }
                 }
+                return true;
+            } else {
+                return false;
             }
-            return false;
         }
 
         @Nullable
         private WorkerId getWorkerId() {
```

Comment:
nit: optional?

Reply:
not required I think given it's explicitly marked as nullable and is particularly used within the class.